### DISCUSSION TITLE
Close completion with Escape key, but remain in Insert mode?

### DISCUSSION THREAD

**Discussion by pbar1** [hrsh7th/nvim-cmp#924] [Q&A] [answered]
I'd like to be able to close nvim-cmp's completion menu with `<Esc>`, but remain in Insert mode. With my current configuration (posted below), pressing `<Esc>` closes completion and exits back to Normal mode. I've also tried adding a couple of different mappings, but neither worked. 

Is this possible?

Current `mapping` block. Here, nothing is bound to `<Esc>`, so I wouldn't expect it to work yet:
```lua
 mapping = {
    ["<C-Space>"] = cmp.mapping.complete(),
    ["<C-e>"] = cmp.mapping.close(),
    ["<CR>"] = cmp.mapping.confirm({
       behavior = cmp.ConfirmBehavior.Replace,
       select = true,
    }),
    ["<Tab>"] = function(fallback)
       if cmp.visible() then
          cmp.select_next_item()
       elseif luasnip.expand_or_jumpable() then
          luasnip.expand_or_jump()
       else
          fallback()
       end
    end,
    ["<S-Tab>"] = function(fallback)
       if cmp.visible() then
          cmp.select_prev_item()
       elseif luasnip.jumpable(-1) then
          luasnip.jump(-1)
       else
          fallback()
       end
    end,
 }
```

Tried adding this to the above (didn't work, exited Insert mode):
```lua
["<Esc>"] = cmp.mapping.abort(),
```

And this (didn't work, still exited Insert mode):
```lua
["<Esc>"] = function(fallback)
   if cmp.visible() then
      cmp.mapping.close()
   else
      fallback()
   end
end,
```

I kinda thought this last one was gonna work, but looking back at it, I'm just calling a mapping on that line. However, to debug things, I changed `cmp.mapping.close()` on that line to `cmp.select_next_item()`, and tried it again. What I hoped would happen here was: In a completion menu, I would press `<Esc>` and the next item would be selected. However, even with this config, I exited Insert mode upon pressing `<Esc>`.

**Comment by arsham** [comment]
Would `<C-e>` be a good option?

**Reply by pbar1** [reply]
That's currently what I've got, but looking for the `<Esc>` mapping in addition to that

**Comment by harrygallagher4** [comment, marked as answer]
Do your `<tab>`/`<s-tab>` maps work? I think functions need to be wrapped in `cmp.mapping(fn)` to work properly.

My configuration is in fennel but I've used the following for months:
```fnl
(local m cmp.mapping)
(local visible? cmp.visible)
(local abort! cmp.abort)

{:mapping {:<esc> (m #(if (visible?) (abort!) ($)) [:i :c])}}
```

The function is basically identical to the last one in your post except it uses abort instead of close.

**Reply by pbar1** [reply]
Yep, `<tab>`/`<s-tab>` do work. I'll try that out - does it essentially translate to this?

```lua
["<Esc>"] = function()
   if cmp.visible() then
      cmp.mapping.abort()
   end
end,
```

Ie, no fallback?

**Reply by harrygallagher4** [reply]
Actually, the `($)` is the fallback. My esc mapping is basically this in lua:

```lua
{
["<esc>"] = cmp.mapping(function(fallback)
  if cmp.visible() then
    cmp.abort()
  else
    fallback()
  end
end, {'i', 'c'})
}
```

p.s.: I believe that *inside* of custom function mappings you shouldn't use `cmp.mapping.abort()` & others, you should instead directly use `cmp.abort()`

**Reply by pbar1** [reply]
This works - thank you! I believe you were right about wrapping this portion inside `cmp.mapping()` - though I'm now unsure how Tab/S-Tab are working without being wrapped this way. But this is great 💯

**Comment by DrKJeff16** [comment]
> Actually, the `($)` is the fallback. My esc mapping is basically this in lua:
> 
> ```lua
> {
> ["<esc>"] = cmp.mapping(function(fallback)
>   if cmp.visible() then
>     cmp.abort()
>   else
>     fallback()
>   end
> end, {'i', 'c'})
> }
> ```
> 
> p.s.: I believe that _inside_ of custom function mappings you shouldn't use `cmp.mapping.abort()` & others, you should instead directly use `cmp.abort()`

Yes, avoid `cmp.mapping.abort()` please. I got stuck in INS mode while editing my lua modules for cmp so I had to kill it externally.
**_INSERT "HOW TO GET OUT OF VIM" JOKE HERE_**

**Comment by hinell** [comment]
@pbar1 You can use [nvim_feedkeys](https://neovim.io/doc/user/api.html#nvim_feedkeys()) for that.
